doc/smilint.1
.\"
.\" $Id: smilint.1,v 1.2 1999/05/05 15:48:19 strauss Exp $
.\"
.TH smilint 1  "May 5, 1999" "IBR" "SMI Tools"
.SH NAME
smilint \- syntax and semantics checks of SMIv1/v2/ng modules
.SH SYNOPSIS
.B smilint
[
.B "-sSvV"
] [
.BI "-l " level
] [
.BI "-d " level
] [
.BI "-c " config
] [
.BI "-L " location
]
.I "module"
.SH DESCRIPTION
The \fBsmilint\fP program is used to check MIB modules for syntax
errors and semantics at some degree.  Both, SMIv1/v2 and SMIng style
MIB modules are supported.
.PP
The rules that \fBsmilint\fP is based on, are taken from RFC 1155, RFC
1212 and RFC 1215 for SMIv1, RFCs 2578-2580 for SMIv2 and the
specifications of SMIng that are part of a master's thesis at the
Technical University, Braunschweig, Germany.
.PP
\fBSmilint\fP first reads the global libsmi configuration file
\fB/usr/local/etc/smi.conf\fP, as any libsmi application does. Then
the command line options are interpreted and finally, the
\fB/usr/local/etc/smilint.conf\fP configuration file is read, if present,
before \fBsmilint\fP starts its main operation.
.SH OPTIONS
.TP
.B "-s"
Report some per module statistics to stderr.
.TP
.B "-S"
Don't report statistics (default).
.TP
.B "-v"
On errors, do print the current line of input to stderr after the
error message and a marker pointing to the current token in the input
stream. Note that the corresponding point of error usually lies before
this marker.
.TP
.B "-V"
Don't print error lines (default).
.TP
.BI "-d " level
Print libsmi debugging messages at the given \fIlevel\fP (default: 0).
.TP
.BI "-l " level
Report errors and warnings up to the given \fIlevel\fP to stderr.
A higher level shows more warnings. The actual levels are:
.sp
.nf
 0   internal error, no recovery possible,
 1   major SMI error, recovery somehow possible but
     may lead to severe problems,
 2   SMI error, probably tolerated by some
     implementations,
 3   SMI error, tolerated by many implementations,
 4   something which is recommended to be changed,
 5   something that might be ok, but might not be
     recommended under some circumstances.
.fi
.sp
Higher levels are actually not used and lead to the same effects as
level 5 does. Note that errors up to level 3 are errors violating the
specification and must be fixed by the responsible author. Level 3 is
the default level.
.TP
.BI "-c " config
Read the configuration from \fIfile\fP. See smi_config(3) section
CONFIGURATION FILES for its format.
.TP
.BI "-L " location
Add \fIlocation\fP to the list of locations to search for MIB modules.
See smi_config(3) section MODULE LOCATIONS for location details.
.TP
.I module
This is the module to be checked. Note that this module is searched in
the list of known locations. If the module shall be taken from an
SMIv1/v2 file in the current directory, a "\fB-L smi:.\fP" option
should be added to the command line.
.SH "FILES"
.nf
/usr/local/etc/smi.conf     general configuration file
/usr/local/etc/smilint.conf smilint configuration file
.fi
.SH "NOTES"
Libsmi and this tool are work in progress. Expect changes in near
future.
.SH "SEE ALSO"
The SMIng and
.BR libsmi (3)
project is documented at
.BR "http://www.ibr.cs.tu-bs.de/~strauss/sming/" "."
Other commonly used MIB checkers are
.BR mosy "(1), "
.BR smicng "(1)."
.SH "AUTHOR"
(C) 1999 Frank Strauss, TU Braunschweig, Germany <strauss@ibr.cs.tu-bs.de>
.br


ChangeLog
2000-11-24  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* libsmi-0.2.10.tar.gz: released version 0.2.10.

	* configure.in: Fixed broken 64 bit type configuration.

2000-11-20  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* changed filenames of file created by the corba and netsnmp
	dump drivers. Adapted test suite.

	* libsmi-0.2.9.tar.gz: released version 0.2.9.

2000-11-16  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* smi.conf-example: Added a configuration file example.

	* configure.in: Dir separator is now configurable and distinguished
	in util/smiIsPath().

	* win/*: Some changes for clean MSVC compilation. Yigal.

2000-11-14  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* win/config.h: Added macros for 64 bit types on MSVC systems.
	Added a necessary compiler flag to compile scanner-*.c on MSVC.
	Thanks to Yigal.

	* configure.in: Fixed: completely broken AC_TRY_RUN check for
	64 bit types. Somehow, I did it in a way that was syntactically
	correct. ;-)

	* acconfig.h: renamed MIN/MAX macros to suppress redefinitions
	on Solaris 2.6+.

2000-11-13  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* libsmi-0.2.8.tar.gz: released version 0.2.8.

	* test/Makefile.am: Added tests for formats: identifiers,
	metrics, xml, cm, python.
	
	* tools/dump-jax.c: Fixed dangling pointers. Test suite
	is happy again.

	* tools/dump-corba.c: The driver now creates idl and oid files
	instead of writing to stdout.
	
	* test/*: Reorganized the test suite. All scripts for
	smidump scripts are now identical. Added jax test (and found
	a bug that is not yet fixed).

2000-11-12  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* tools/dump-netsnmp.c: Renamed dump-ucdsnmp.c to dump-netsnmp.c.

	* test/Makefile.am: Commented out checks for the obsolete
	corba-{idl,oid} checks to make the test suite happy.

	* configure.in: The availability of `long long' type is now
	checked during configuration. The result now also affects printf
	formats.

	* configure.in: Path separator is now configurable (useful to
	use ":" instead of ";" as well in cygwin environments).

2000-11-09  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* tools/smidump.c: Juergen did a major rework of the internal
	smidump interface for the output drivers. Visible changes:
	corba-idl and corba-oid are integrated into a single format
	corba; -o option.

2000-11-08  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/iana/*: Updated IANA maintained modules.

	* lib/parser-smi.y: Suppressed an module conformance error
	message in case of SMIv1.

2000-11-08  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/check.c,parser-smi.y: Bill Fenner fixed some duplicated
	error messages.

	* tools/dump-smi.c,dump-tree.c: Bill Fenner fixed some core
	dumps on erroneous MIB modules.

	* tools/dump-ucdsnmp.c: Juergen changed the UCD-SNMP driver
	so that it now creates three file (header, stubs & implementation)
	with a single smidump format option (netsnmp).

	* lib/parser-smi.y: Added checks for illegal subtyping; adjusted
	test suite, so that is runs successfully for all current tests.

	* lib/scanner-smi.l,scanner-sming.l: Bill Fenner contributed
	a patch to continue scanning after unexpected characters have
	been read.

2000-11-06  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* libsmi-0.2.7.tar.gz: released version 0.2.7.

	* lib/parser-smi.y: Fixed `0' ranges for some core types on Solaris.

	* lib/smi_config.3.in: Updated the library man pages.

	* tools/smidump.c: addModule() now inserts NULL module at the
	head of the list as expected by the dump drivers.

	* lib/check.c: Added checks for OID redifinitions and recursive
	definitions, based on a contribution from Bill Fenner.

	* lib/smi.h.in: Added a SmiModule.conformance attribute that
	allows applications to determine, `how syntactically correct'
	a module is.

2000-11-04  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/DISMAN-EVENT-MIB: added (RFC 2981).

	* mibs/ietf/DISMAN-EXPRESSION-MIB: added (RFC 2982). It needed
	a small fix.

2000-11-01  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/smi.c: Reworked the smiPath initialization upon a hint
	from Bill Fenner. Note that the order of evaluation of the SMIPATH
	environment variable and configuration files has changed and that
	the syntax of `path' commands in configuration files has changed.
	See smi_config(3), section MODULE LOCATIONS for details.

2000-10-27  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/check.c: Juergen fixed index checks for Bits/fixed-length
	octet strings.

2000-10-25  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* tools/smidump.c: Removed again `-X' option to eliminate parser
	options that might lead to different results.

2000-10-21  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/error.c: Added a fourth arg to the error handler representing
	the error name. This has been wished and contributed by Bill Fenner.
	Note that this changed the API slightly.

2000-10-20  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/check.c: Fixed index length checks for OCTET STRINGs,
	IpAddresses and OBJECT-IDENTIFIERs. Thanks to Bill Fenner.

	* lib/error.c: Fixed two minor typos. Thanks to Bill.

2000-10-19  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* tools/dump-python.c: a new dump format to generate Python MIB
	dictionary code. Contributed by Pat Knight.

	* tools/dump-xml.c: Fixed missing end tag of rows and tables
	in certain cases. Contributed by Pat Knight.

	* tools/smidump.c: Added option `-X' for `lax' parsing. Some
	people wish to accept things like underscores or upper case
	first letters in object type names. Note that it is usually
	a *bad* idea to use lax parsing!

2000-10-18  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/FRNETSERV-MIB: updated (RFC 2954).

	* mibs/ietf/FR-ATM-PVC-SERVICE-IWF-MIB: added (RFC 2955).

2000-10-17  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* tools/dump-ucdsnmp.c: Juergen fixed BITS handling.

2000-10-12  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/RTP-MIB: added (RFC 2959).


tools/smilint.c
/*
 * smilint.c --
 *
 *      MIB module checker main program.
 *
 * Copyright (c) 1999 Frank Strauss, Technical University of Braunschweig.
 *
 * See the file "COPYING" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * @(#) $Id: smilint.c,v 1.23 2000/02/09 19:56:52 strauss Exp $
 */

#include <config.h>

#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#include "smi.h"



void usage()
{
    fprintf(stderr,
	    "Usage: smilint [-Vhsr] [-c <configfile>] [-p <module>] [-l <level>]\n"
	    "               [-i <error-pattern>] <module_or_path>\n"
	    "-V                    show version and license information\n"
	    "-h                    show usage information\n"
	    "-s                    print statistics on parsed MIB modules\n"
	    "-r                    print errors also for imported modules\n"
	    "-c <configfile>       load a specific configuration file\n"
	    "-p <module>           preload <module>\n"
	    "-l <level>            set maximum level of errors and warnings\n"
	    "-i <error-pattern>    ignore errors matching prefix pattern\n"
	    "<module_or_path>      plain name of MIB module or file path\n");
}



void version()
{
    printf("smilint " VERSION "\n");
}



int main(int argc, char *argv[])
{
    char c;
    int flags, i;

    for (i = 1; i < argc; i++) if (strstr(argv[i], "-c") == argv[i]) break;
    if (i == argc) 
	smiInit("smilint");
    else
	smiInit(NULL);

    flags = smiGetFlags();
    
    flags |= SMI_FLAG_ERRORS;
    smiSetFlags(flags);
    
    while ((c = getopt(argc, argv, "Vhsrp:l:i:c:")) != -1) {
	switch (c) {
	case 'c':
	    smiReadConfig(optarg, "smiquery");
	    break;
	case 'V':
	    version();
	    exit(0);
	case 'h':
	    usage();
	    exit(0);
	case 'p':
	    smiLoadModule(optarg);
	    break;
	case 'i':
	    smiSetSeverity(optarg, 9);
	    break;
	case 'l':
	    smiSetErrorLevel(atoi(optarg));
	    break;
	case 'r':
	    flags |= SMI_FLAG_RECURSIVE;
	    smiSetFlags(flags);
	    break;
	case 's':
	    flags |= SMI_FLAG_STATS;
	    smiSetFlags(flags);
	    break;
	default:
	    usage();
	    exit(1);
	}
    }

    while (optind < argc) {
	if (smiLoadModule(argv[optind]) == NULL) {
	    fprintf(stderr, "smilint: cannot locate module `%s'\n",
		    argv[optind]);
	    exit(1);
	}
	optind++;
    }

    smiExit();
    
    exit(0);
}
